domain/upgrade/service/watcher.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/errors"
+	"github.com/juju/worker/v3/catacomb"
+
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/juju/core/watcher"
+)
+
+type upgradeReadyWatcher struct {
+	ctx      context.Context
+	catacomb catacomb.Catacomb
+
+	st          State
+	wf          WatcherFactory
+	upgradeUUID string
+
+	in  <-chan []string
+	out chan struct{}
+}
+
+// NewUpgradeReadyWatcher creates a watcher which notifies when all controller
+// nodes have been registered, meaning the upgrade is ready to start
+func NewUpgradeReadyWatcher(ctx context.Context, st State, wf WatcherFactory, upgradeUUID string) (watcher.NotifyWatcher, error) {
+	w := &upgradeReadyWatcher{
+		ctx:         ctx,
+		st:          st,
+		wf:          wf,
+		upgradeUUID: upgradeUUID,
+		out:         make(chan struct{}),
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return w, nil
+}
+
+func (w *upgradeReadyWatcher) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+func (w *upgradeReadyWatcher) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *upgradeReadyWatcher) Changes() <-chan struct{} {
+	return w.out
+}
+
+func (w *upgradeReadyWatcher) loop() error {
+	defer close(w.out)
+
+	namespaceWatcher, err := w.wf.NewNamespaceWatcher("upgrade_info_controller_node", changestream.Create|changestream.Update, "")
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := w.catacomb.Add(namespaceWatcher); err != nil {
+		return errors.Trace(err)
+	}
+	w.in = namespaceWatcher.Changes()
```

Comment:
Unlike `w.out`, this is only used in the loop. You don't actually need a member for it.
Just assign the local `in` to either `namespaceWatcher.Changes()` or `nil` based on tick/tock.

Reply:
Ah yes. This redundancy is an artifact from earlier code I believe